analysis/src/CompletionBackEnd.ml

```diff
@@ -1757,53 +1803,33 @@ Note: The `@react.component` decorator requires the react-jsx config to be set i
-          | Variant constructors ->
-            if debug then
-              Printf.printf "Found variant type for NamedArg typed context %s\n"
-                (typeExpr |> Shared.typeToString);
-            (* TODO: Investigate completing seen identifiers _with the correct type_ here too. If completing
-               variant someVariant, and there's a someVariable of type someVariant, add that to the completion list. *)
-            constructors
-            |> List.filter (fun constructor ->
-                   Utils.startsWith constructor.Constructor.cname.txt prefix)
-            |> List.map (fun (constructor : Constructor.t) ->
-                   (* TODO: Can we leverage snippets here for automatically moving the cursor when there are multiple payloads?
-                      Eg. Some($1) as completion item. *)
-                   Completion.create
-                     ~name:
-                       (constructor.cname.txt
-                       ^
-                       if constructor.args |> List.length > 0 then "(_)" else ""
-                       )
-                     ~kind:(Constructor (constructor, ""))
-                     ~env)
-          | _ -> [])
-        | Some (Polyvariant (_env, constructors)) ->
-          if debug then
-            Printf.printf
-              "Found polyvariant type for NamedArg typed context: %s\n"
-              (typeExpr |> Shared.typeToString);
-          (* TODO: Investigate completing seen identifiers _with the correct type_ here too. Unsure if that's possible though
-              for polyvariants, if things aren't referred to by typename, but rather the full inlined type definition.. *)
-          constructors
-          |> List.filter (fun constructor ->
-                 Utils.startsWith constructor.name prefix)
-          |> List.map (fun constructor ->
-                 (* TODO: Can we leverage snippets here for automatically moving the cursor when there are multiple payloads?
-                    Eg. Some($1) as completion item. *)
-                 Completion.create
-                   ~name:
-                     ("#" ^ constructor.name
-                     ^
-                     if constructor.payload |> Option.is_some then "(_)" else ""
-                     )
-                   ~kind:(PolyvariantConstructor constructor) ~env))))
+      | Some (_, typeExpr) ->
+        (* TODO: Wonder if we can filter the values returned here on type...
+           Maybe that'd be confusing for the user though, if items in scope don't appear.
+           We should at a minimum be able to sort values according to type and the proximity
+           of them in the scope. *)
+        let completionsFromContext =
+          contextPath
+          |> getCompletionsForContextPath ~package ~opens ~rawOpens ~allFiles
+               ~pos ~env ~exact:forHover ~scope
+        in
+        completeTypeExpr ~env ~package ~debug ~prefix typeExpr
+        @ completionsFromContext)
```

Comment:
Sorry, this diff got really messy.

Correct, `completeTypeExpr` is the refactoring and `completionsFromContext` is the change. 

The concatenation is intentional. My reasoning is I need to bring in "general" completions too, otherwise I wouldn't complete for the following in the "typed" scenario:
* Variables you might have in scope that has the value
* Other modules you might want to use something from to get what you need.

Makes sense?

Reply:
If the type is known (and interesting, e.g. a variant), there's no need to complete based on scope unless I'm missing something.
Do you have an example of where this is not the case?

Reply:
We might be talking about two separate things here. Here are two examples of what I mean:

```rescript
type someVariant = One | Two

let someValue = One

let someFunc = (~value: someVariant) => {...

// Here what I want is to use `someValue`. That's a local value I need to bring into the completion, or there's no way for me to access it via completing.
someFunc(~value=
```

And same as above, but with the value in another module/file.
```rescript
// Foo.res
type someVariant = One | Two

let someValue = One

let someFunc = (~value: someVariant) => {...

// Bar.res
// Here what I want is to use `Foo.someValue`. So I need to be able to complete for `Foo`, so I then can complete for `someValue`.
Foo.someFunc(~value=
```

Are you following what I mean?

Reply:
You should not care where it is defined. The type definition gives a list of variants and you use them.

Reply:
There are no general completions in those examples, as neither uses local bindings. They both use the type instead.
Local bindings are not restricted to variants of that type, but will return anything that is in scope.

I interpret that list append as: things from the type + everything in scope.